File: 1/soln.cpp
#include <iostream>
#include <string>
#include <vector>
#include <array>
#include <regex>

std::vector<int> extractDigits(const std::string &str) {
	std::vector<int> digits;

	for (char ch : str) {
		if (ch >= '0' && ch <= '9') {
			digits.push_back(ch - '0');
		}
	}

	return digits;
}

std::vector<int> extractDigitStrings(const std::string &str) {
	std::array<std::string, 10> digitStrings = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };

	std::regex digitRegex("(?=(\\d|one|two|three|four|five|six|seven|eight|nine)).");

	std::vector<int> digits;

	std::sregex_iterator next(str.begin(), str.end(), digitRegex);
	std::sregex_iterator end;

	while (next != end) {
		std::smatch match = *next;
		// std::cout << match.str(1) << "\t" << "\t" << match.position() << "\t" << "\n";

		std::string matchStr = match.str(1);

		if ('0' <= matchStr[0] && matchStr[0] <= '9') {
			digits.push_back(matchStr[0] - '0');
		} else {
			const std::string *s = std::find(digitStrings.begin(), digitStrings.end(), matchStr);
			int n = std::distance(digitStrings.cbegin(), s);

			digits.push_back(n);
		}

		// std::cout << digits.back() << '\n';

		next++;
	}

	// std::cout << '\n';

	return digits;
}

int main() {
	std::string line;
	std::vector<std::string> lines;

	while (std::getline(std::cin, line)) {
		lines.push_back(line);
	}

	std::cout << "Part 1\n";
	std::cout << "\n";

	int sum1 = 0;
	for (const std::string &line : lines) {
		std::vector<int> digits = extractDigits(line);

		if (digits.size() > 0) {
			int value = 10 * digits.front() + digits.back();

			sum1 += value;
		}
	}

	std::cout << sum1 << std::endl;

	std::cout << "\n";
	std::cout << "Part 2\n";
	std::cout << "\n";

	int sum2 = 0;
	for (const std::string &line : lines) {
		std::vector<int> digits = extractDigitStrings(line);

		if (digits.size() > 0) {
			int value = 10 * digits[0] + digits.back();

			sum2 += value;
		}
	}

	std::cout << sum2 << std::endl;
}

File: 1/soln.py
import sys
import re


def extract_digits(string):
    return [int(ch) for ch in string if ch.isdigit()]


def extract_digit_strings(string):
    digit_strings = ['one', 'two', 'three', 'four',
                     'five', 'six', 'seven', 'eight', 'nine']

    regex = r'(?=(\d|' + '|'.join(digit_strings) + r'))'

    # print(regex)

    return [int(m.group(1)) if m.group(1).isdigit() else digit_strings.index(
        m.group(1))+1 for m in re.finditer(regex, string)]


def cat_first_last(digits):
    if digits == None or len(digits) == 0:
        return 0

    return 10 * digits[0] + digits[-1]


lines = [line.strip() for line in sys.stdin]

print('Part 1')
print(sum(cat_first_last(extract_digits(l)) for l in lines))

print('Part 2')
print(sum(cat_first_last(extract_digit_strings(l)) for l in lines))
